models/biocro/R/run.biocro.R

```diff
@@ -51,111 +44,177 @@ run.biocro <- function(lat, lon, metpath, soil.nc = NULL, config = config, coppi
       day1 <- NULL
       dayn <- NULL
     }
+    if (!is.null(day1) && day1 > lubridate::yday(starti)) {
+      lubridate::yday(starti) <- day1
+    }
+    if (!is.null(dayn) && dayn < lubridate::yday(starti)) {
+      lubridate::yday(plant_date) <- dayn
+    }
+    WetDat <- WetDat[WetDat$doy >= lubridate::yday(starti) & WetDat$doy <= lubridate::yday(endi), ]
 
-    # BLETCHEROUS HACK: BioCro 0.94 starts the run by subsetting weather data
-    # to day1:dayn, but it assumes the data start on DOY 1 and contain
-    # (yearlength*(24/timestep)) lines. This means that in practice, day1 and
-    # dayn are treated as "day of file" not "day of year".
-    # BioCro *does* handle DOY correctly downstream of the subsetting, so here
-    # we check if the current BioCro has fixed this assumption.
-    # If not, rescale day1 and dayn to be relative to the start of the input.
-    #   Scaling is derived by inverting Biocro's day->index equations.
-    biocro_checks_doy <- tryCatch(
-      {m <- BioCro::BioGro(WetDat = matrix(c(0,10,0,0,0,0,0,0),nrow = 1),
-                   day1 = 10, dayn = 10, timestep = 24);
-      class(m) == "BioGro"},
-      error = function(e){FALSE})
-    if (!biocro_checks_doy && min(WetDat[,"doy"])>1) {
-      if (!is.null(day1)){
-        # Biocro calculates line number as `indes1 <- (day1 - 1) * 24`
-        indes1 <- Position(function(x)x==day1, WetDat[,"doy"])
-        day1 <- indes1/24 + 1
+    HarvestedYield <- 0
+
+    if (utils::packageVersion('BioCro') >= 1.0) {
+
+      if ("SolarR" %in% names(WetDat)) {
```

Comment:
@justinmcgrath should this move to BioCro v1 so that it accepts ( `solar` or `SolarR`)  and (`WS` or `windspeed` ), allowing backward compatibility w/ existing weather datasets? (or does it already accept these synonyms?)

Reply:
It will break if the wrong names are given, which is in line with the single responsibility principle. I think this is an issue because the weather data sets in BioCro do not use consistent names, so we could unify the names and simplify the code, which I'd quite like to do. If the preferred names in the data sets that Pecan uses are 'WS' and 'SolarR', and thus they cannot be renamed in the data set, then Pecan is the appropriate place for the name change, since it is the only part of the code that should know about both where the parameters come from and where they will be used. This particular spot does not seem like a great place for it though, but it's not immediately obvious to me what place would be better.

As a general solution to this problem, I would define sets of mapping between names that have the same meaning. For any new model that uses different names, provide a mapping to a set of standardized names. Then it's straightforward to write a function that performs the conversion as follows:
parameters_for_b = read(parameter_set, as='set_a').as('set_b')

The mapping could also include conversions for units. The mapping has the attractive feature of being self documenting. As long as a consistent format is used, the mapping could be provided anywhere: in the original package, or elsewhere.

Reply:
> If the preferred names in the data sets that Pecan uses are 'WS' and 'SolarR', 

PEcAn doesn't care what the names are, and already has a general solution to this problem: each model wrapper has a function analogous to met2model.BIOCRO function that handles name and unit conversions with varying degrees of elegance. 

I forgot that BioCro v0.9x didn't care what the column names are. In that case, we can make the met2model.BIOCRO output provide BioCro v1.0 names (solar and windspeed) and remove this code since v0.9x won't care.

Reply:
I think that v0.9 does use the column names in some places and indexes in other places - rather annoying. I know I changed some of the indexes to column names, but I know it didn't change everything because it was easier to just get rid of that stuff entirely.